### PR TITLE
feat: web push API

### PR DESCRIPTION
**Pull request by caalador**
Add API to use initialize web push on the client from the server.

Related-to https://github.com/vaadin/platform/issues/4037

Fixes https://github.com/vaadin/flow/issues/14606

### CODE REVIEW

**Comment by mshabarov** [comment]
`WebPush` can be created explicitly as `WebPush webPush = new WebPush(keys)` and then for DI frameworks it can be made as a singleton managed bean, e.g. for Spring:

```java
    @Bean(name = "WebPush")
    public WebPush getWebPush() { 
        // keys are injected with AtValue
        return new WebPush(keys);
    }
```

**Comment by mshabarov** [comment]
Pushed an extension for `WebPushRegistration` regarding sending notifications.

Also updated base web push example here https://gitlab.vaadin.com/flow/base-webpush/-/tree/send-notification-api?ref_type=heads to showcase the API.

File: flow-server/src/main/java/com/vaadin/flow/server/frontend/webpush/WebPushRegistration.java [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+     *
+     * @param subscription
+     *            web push subscription of the client
+     * @param messageJson
+     *            notification message containing <code>title</code> and
+     *            <code>body</code> strings
+     * @throws WebPushException
+     *             if sending a notification fails
+     */
+    public abstract void sendNotification(
```
Comment (mshabarov) [inline flow-server/src/main/java/com/vaadin/flow/server/frontend/webpush/WebPushRegistration.java:159]: We can create an entity class for message (restricted to only title and body at this point) and give it as a parameter here instead. WDYT?

**Review by mshabarov — COMMENTED** [review]
Have some further suggestions, let's discuss.

File: flow-server/src/main/java/com/vaadin/flow/server/PwaIcon.java [hunk 2]
```diff
@@ -16,6 +16,7 @@
 package com.vaadin.flow.server;
 
 import javax.imageio.ImageIO;
+import java.awt.*;
```
[the remark this answers is not in the document]
Comment (caalador) [reply flow-server/src/main/java/com/vaadin/flow/server/PwaIcon.java:19]: fixed.
Issue of new IDE with wrong settings.

File: flow-server/src/main/java/com/vaadin/flow/server/PwaIcon.java [hunk 3]
```diff
@@ -243,4 +255,43 @@ public void write(OutputStream outputStream) {
         }
     }
 
+    private BufferedImage drawIconImage(BufferedImage baseImage) {
```
[the remark this answers is not in the document]
Comment (caalador) [reply flow-server/src/main/java/com/vaadin/flow/server/PwaIcon.java:260]: it fixes the slowness of adding PWA as it generated on start all images and now we only generate images on request instead.

[the remark this answers is not in the document]
Comment (caalador) [reply flow-server/src/main/java/com/vaadin/flow/server/PwaIcon.java:260]: see 14606

[the remark this answers is not in the document]
Comment (mshabarov) [reply flow-server/src/main/java/com/vaadin/flow/server/PwaIcon.java:260]: Thanks, good to have it. We can reference this issue in the commit description.

**Comment by mshabarov** [comment]
Looking into possible further improvements of Web Push API (outside of this pull request), I made a short analysis of what we might need in the future:

- Notifications might have a link to give users opportunity to navigate to a particular page/content on the web site/mobile app relevant to this notification, so when the user clicks on the notification, browser redirects to a page containing relevant content.
    - discussed here https://stackoverflow.com/questions/43986805/open-a-custom-url-when-clicking-on-a-web-push-notification, need a new parameter for link in the message object passed to `sendNotification` method.

- Updating or removing pending notifications, so that user can see one notification with: "You have 3 unread messages". The ability to replace a notification: replace "You have 3 unread messages" with "You have 7 unread messages" instead of adding a new notification for each message.
    - doesn't need a new Java API, but could be an implementation change

- Dropdown component shown on the notification body with customisable list of clickable links.
    - Java API needs a parameter containing a list of links

- API for seeing whether sending notification is available for the user (feature detection in browsers, allowed or blocked by user)
    - add a new Java API (methods) implementing calls to JavaScript permission API, see https://developer.mozilla.org/en-US/docs/Web/API/Permissions_API

- The ability to clear the pending notifications in browser
    - add a new Java API (method) implementing call to a JavaScript `close` method, see https://developer.mozilla.org/en-US/docs/Web/API/Notification/close

So it seems like we'll need a new methods in `WebPushRegistration` or new fields in the `Message` object, and this can be added iteratively.

**Comment by caalador** [comment]
Permission check is avaliable in the registration script, but doesn't have a server method.

File: flow-server/src/main/java/com/vaadin/flow/server/frontend/webpush/WebPushRegistration.java [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+
+    private final SerializableConsumer<String> errorHandler = err -> {
+        throw new RuntimeException("Unable to retrieve extended "
+                + "client details. JS error is '" + err + "'");
+    };
+
+    /**
+     * Web push subscription class containing web push registration data.
+     */
+    public static class WebPushSubscription {
```
[the remark this answers is not in the document]
Comment (caalador) [reply flow-server/src/main/java/com/vaadin/flow/server/frontend/webpush/WebPushRegistration.java]: done

File: flow-server/src/main/java/com/vaadin/flow/server/frontend/webpush/WebPushRegistration.java (hunk 1 above)
Comment (mshabarov) [reply flow-server/src/main/java/com/vaadin/flow/server/frontend/webpush/WebPushRegistration.java]: Added

File: flow-server/src/main/java/com/vaadin/flow/server/frontend/webpush/WebPushRegistration.java [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+import com.vaadin.flow.component.dependency.JsModule;
+import com.vaadin.flow.function.SerializableConsumer;
+
+import elemental.json.Json;
+import elemental.json.JsonObject;
+import elemental.json.JsonValue;
+
+/**
+ * Class for handling web push registration to the client and returning
+ * subscription data on the server for pushing notifications.
```
[the remark this answers is not in the document]
Comment (mshabarov) [reply flow-server/src/main/java/com/vaadin/flow/server/frontend/webpush/WebPushRegistration.java]: Added

**Comment by mshabarov** [comment]
This is how the Web Push API looks like now:

WebPush service implementation:
```java
@Service
@ApplicationScope
public class WebPushService extends WebPushRegistration {

    private final PushService pushService;

    private final Map<String, WebPushSubscription> endpointToSubscription =
            new HashMap<>();

    /**
     * Create new WebPushRegistration for given publicKey.
     *
     * @param publicKey public key to use for web push
     */
    public WebPushService(@Value("${public.key}") String publicKey,
                          @Value("${private.key}") String privateKey,
                          @Value("${subject}") String subject) throws WebPushException {
        super(publicKey);
        Security.addProvider(new BouncyCastleProvider());
        // Initialize push service with the public key, private key and subject
        try {
            pushService = new PushService(publicKey, privateKey, subject);
        } catch (GeneralSecurityException e) {
            throw new WebPushException("Failed to initialise push service", e);
        }
    }

    @Override
    public void sendNotification(WebPushSubscription subscription,
                                 WebPushMessage message) throws WebPushException {
        try {
            Subscription pushServiceSubscription = new Subscription(
                    subscription.getEndpoint(),
                    new Subscription.Keys(subscription.getP256dh(),
                            subscription.getAuth()));
            pushService.send(new Notification(pushServiceSubscription,
                            message.toJson()));
        } catch (GeneralSecurityException | IOException | JoseException | ExecutionException
                 | InterruptedException e) {
            throw new WebPushException("Failed to send a message", e);
        }
    }

    public void subscribe(UI ui) {
        subscribeWebPush(ui, subscription -> endpointToSubscription.put(
                subscription.getEndpoint(), subscription));
    }

    public void unsubscribe(UI ui) {
        unsubscribeWebPush(ui, subscription -> endpointToSubscription.remove(
                subscription.getEndpoint()));
    }

    /**
     * Send a notification to all subscriptions.
     *
     * @param message web push message with title and body
     */
    public void notifyAll(WebPushMessage message) {
        endpointToSubscription.values().forEach(subscription -> {
            try {
                sendNotification(subscription, message);
            } catch (WebPushException e) {
                getLogger().error("Failed to send a message", e);
            }
        });
    }

    private Logger getLogger() {
        return LoggerFactory.getLogger(WebPushService.class);
    }

}
```

User's view
```java
@Route(value = "user", layout = MainLayout.class)
public class UserView extends VerticalLayout {

    private final Checkbox toggle;

    private final WebPushService webPushService;

    public HelloWorldView(WebPushService webPushService) {
        Paragraph paragraph = new Paragraph("Subscribe for Web Push Notifications to receive " +
                                            "real-time updates, reminders, and other important messages");
        this.webPushService = webPushService;
        toggle = new Checkbox("Subscribed");
        add(paragraph, toggle);
    }

    @Override
    protected void onAttach(AttachEvent attachEvent) {
        super.onAttach(attachEvent);
        UI ui = attachEvent.getUI();
        webPushService.isWebPushRegistered(ui, toggle::setValue);
        // omitted listener removal on detach
        toggle.addValueChangeListener(event -> {
            if (event.getValue()) {
                webPushService.subscribe(ui);
            } else {
                webPushService.unsubscribe(ui);
            }
        });
    }
}
```

And admin view from where the notification can be triggered:
```java
@Route(value = "admin", layout = MainLayout.class)
public class AdminView extends HorizontalLayout {

    private TextField message;
    private Button broadcast;

    public AdminView(WebPushService webPushService) {
        message = new TextField("Message");
        broadcast = new Button("Broadcast message", e ->
                webPushService.notifyAll(new WebPushMessage("Message from user",
                        message.getValue()))
        );
        broadcast.addClickShortcut(Key.ENTER);
        add(message, broadcast);
    }
}
```

**Comment by mshabarov** [comment]
We discussed API and implementation here are comments to be addressed:
- include `PushService` implementation and whole library to the Flow
- all Web Push related classes can be extracted into a separate module `flow-web-push`
- `WebPushRegistration` to be renamed to `WebPush`
- Subscribe method can take two callbacks and unsubscribe method can take only UI, like shown below:
```java
            if (subscribed) {
                webPush.subscribe(ui,
                        subscription -> subscriptionService.add(subscription.getEndpoint(),
                                subscription),
                        subscription -> subscriptionService.remove(subscription));
            } else {
                webPush.unsubscribe(ui);
            }
```
- Think about method name `isWebPushRegistered` to better reflect what does this method do. Probably we should avoid using "registration" in the methods/class names, and use "subscription" instead.

File: flow-webpush/src/main/java/com/vaadin/flow/server/webpush/WebPush.java [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+ * <p>
+ * This class doesn't include any implementation for the intercommunication with
+ * a third-party Push Server. Abstract method
+ * {@link #sendNotification(Subscription, WebPushMessage)} to be extended by
+ * developers to use a concrete implementation/library, which sends a
+ * notification to Push Server and later on to a browser.
+ *
+ * @since 24.2
+ */
+@JsModule("./FlowWebPush.js")
```
Comment (Artur-) [inline flow-webpush/src/main/java/com/vaadin/flow/server/webpush/WebPush.java:59]: Does this actually work? `WebPush` is not a component

Comment (caalador) [reply flow-webpush/src/main/java/com/vaadin/flow/server/webpush/WebPush.java:59]: at least for devmode it does... Haven't started the application in production mode.

Comment (caalador) [reply flow-webpush/src/main/java/com/vaadin/flow/server/webpush/WebPush.java:59]: Works fine in production mode also as the class is found as used.

Comment (Artur-) [reply flow-webpush/src/main/java/com/vaadin/flow/server/webpush/WebPush.java:59]: If it for some reason ends up in a non-eager bundle, it won't work

File: flow-webpush/src/main/java/com/vaadin/flow/server/webpush/WebPush.java [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+     *             if sending a notification fails
+     */
+    public void sendNotification(Subscription subscription,
+            WebPushMessage message) throws WebPushException {
+        try {
+            HttpResponse response = pushService
+                    .send(new Notification(subscription, message.toJson()));
+            int statusCode = response.getStatusLine().getStatusCode();
+            if (statusCode != 201) {
+                getLogger().error("Server error, status code:" + statusCode);
```
Comment (mshabarov) [inline flow-webpush/src/main/java/com/vaadin/flow/server/webpush/WebPush.java:126]: This message can be more detailed on what is failed, like "Web Push Notification sending failed, status code returned by Push Server: "

**Review by mshabarov — CHANGES_REQUESTED** [review]
Some minor comments about javadoc and feature flags.
About testing: unit tests are not much interesting for this feature, because half of the feature is in the browser.
We need an end-to-end test, but need a discussion how properly set up the testing environment.

File: flow-webpush/src/main/java/com/vaadin/flow/server/webpush/WebPush.java [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+import elemental.json.JsonValue;
+
+/**
+ * Base class for handling Web Push notifications.
+ * <p>
+ * Enables developers to register clients to the Push Server, return
+ * subscription data to be stored on a server, unregister clients and sending
+ * notifications to the clients.
+ * <p>
+ * This class doesn't include any implementation for the intercommunication with
```
[the remark this answers is not in the document]
Comment (caalador) [reply flow-webpush/src/main/java/com/vaadin/flow/server/webpush/WebPush.java]: removed

File: flow-webpush/src/main/java/com/vaadin/flow/server/webpush/WebPush.java [hunk 9]
```diff
@@ -0,0 +1,280 @@
...
+     * @param privateKey
+     *            web push private key
+     * @param subject
+     *            Subject used in the JWT payload (for VAPID).
+     */
+    public WebPush(String publicKey, String privateKey, String subject) {
+        if (!FeatureFlags.get(VaadinService.getCurrent().getContext())
+                .isEnabled(FeatureFlags.WEB_PUSH)) {
+            getLogger().error(
+                    "WebPush feature is not enabled. Enable feature though dev window or feature file.");
```
[the remark this answers is not in the document]
Comment (caalador) [reply flow-webpush/src/main/java/com/vaadin/flow/server/webpush/WebPush.java]: Updated.
The problem there is that then you can't use the dev tools window to enable the feature.

File: flow-webpush/src/main/java/com/vaadin/flow/server/webpush/WebPush.java (hunk 7 above)
Comment (caalador) [reply flow-webpush/src/main/java/com/vaadin/flow/server/webpush/WebPush.java]: Updated log message

File: flow-webpush/src/main/java/com/vaadin/flow/server/webpush/WebPush.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (mshabarov) [reply flow-webpush/src/main/java/com/vaadin/flow/server/webpush/WebPush.java]: Then maybe not mention dev tools at all here? How this related to throwing the exception? Can we change from logging to throw?

[the remark this answers is not in the document]
Comment (caalador) [reply flow-webpush/src/main/java/com/vaadin/flow/server/webpush/WebPush.java]: We can always throw, but I feel it is more intuitive to add it through the dev tools window than to start figuring out how and where the feature flags file should be.

[the remark this answers is not in the document]
Comment (caalador) [reply flow-webpush/src/main/java/com/vaadin/flow/server/webpush/WebPush.java]: Added throws and message

**Review by mshabarov — APPROVED** [review]

File: flow-webpush/src/main/java/com/vaadin/flow/server/webpush/WebPush.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (mshabarov) [reply flow-webpush/src/main/java/com/vaadin/flow/server/webpush/WebPush.java]: Ok, we can consider making dev tools toggle work later on in a separate PR.

File: flow-webpush/src/main/java/com/vaadin/flow/server/webpush/WebPush.java (hunk 6 above)
Comment (caalador) [reply flow-webpush/src/main/java/com/vaadin/flow/server/webpush/WebPush.java]: Changed to executeJs loading of the scripts on need.

**Pull request merged**